### PR TITLE
OSD-14647 - Reinvestigate clusters that have nodes missing when CHGM resolves

### PR DESCRIPTION
**Pull request by tnierman**
Completes: https://issues.redhat.com/browse/OSD-14647

---

These changes:
- remove an unnecessary check from the `investigateRestoredCluster` function, where recently terminated machines caused CAD's alert resolution investigation to fail
- introduce retry logic into CAD's `evaluateRestoredCluster` function, causing CAD to re-investigate clusters after a short delay. This provides time for them to fully recover from a stopped state, but is only carried out if no errors were encountered in the initial investigation.
- renames the alert generated by CAD when the investigation following CHGM alert resolution fails, based on feedback.

### CODE REVIEW

File: cadctl/cmd/cluster-missing/cluster-missing.go [hunk 1]
```diff
@@ -34,6 +34,8 @@ import (
 
 const (
 	pagerdutyIncidentResolved = "incident.resolved"
+
+	restoredClusterRetryDelay = 600 // seconds
```
[the remark this answers is not in the document]
Comment (tnierman) [reply cadctl/cmd/cluster-missing/cluster-missing.go]: The initial estimation was based on the machinehealthcheck definition for OSD/ROSA [worker](https://github.com/openshift/managed-cluster-config/blob/master/deploy/osd-machine-api/011-machine-api.srep-worker-healthcheck.MachineHealthCheck.yaml#L38-L44) & [infra](https://github.com/openshift/managed-cluster-config/blob/master/deploy/osd-machine-api/010-machine-api.srep-infra-healthcheck.MachineHealthCheck.yaml#L15-L21) machines, which requires their `Ready` status be `false` or `unknown` for 480 seconds (8 minutes) before the machine-api replaces them. The 10 minute value was found to be sufficient for newly stopped machines, and given that CAD is expected to evaluate clusters after they have recovered to a sufficient degree where they can reach Deadmanssnitch, it seemed like a reasonable, if slightly conservative, starting point for the delay period.

However, on re-evaluating this value, it looks like `metal` instance types are permitted a [longer timeout](https://github.com/openshift/managed-cluster-config/blob/master/deploy/osd-machine-api/012-machine-api.srep-metal-worker-healthcheck.MachineHealthCheck.yaml#L42-L44) (15 minutes) when their `Ready` status is `unknown`. In light of this, I've updated CAD's retry delay to 20 minutes, which should be more than sufficient for most clusters.

File: pkg/services/chgm/chgm.go [hunk 2]
```diff
@@ -320,27 +321,17 @@ func (c Client) investigateStartedInstances() (InvestigateInstancesOutput, error
...
-	// Verify cluster has expected number of nodes
+	// Verify cluster has expected number of nodes running
 	nodeCount := c.GetNodeCount()
 	runningInstances, err := c.ListRunningInstances(infraID)
 	if err != nil {
 		return InvestigateInstancesOutput{}, fmt.Errorf("could not retrieve running instances for %s: %w", infraID, err)
 	}
 	if len(runningInstances) != nodeCount {
-		return InvestigateInstancesOutput{UserAuthorized: true, Error: "number of running instances does not match the expected node count: quota may be insufficient or irreplaceable machines have been terminated"}, nil
+		return InvestigateInstancesOutput{UserAuthorized: true, Error: "number of running instances does not match the expected node count: quota may be insufficient or irreplaceable machines have been terminated", Retry: true}, nil
```
[the remark this answers is not in the document]
Comment (tnierman) [reply pkg/services/chgm/chgm.go]: I've updated the logic per your suggestion. 

In my personal opinion, the [original submission](https://github.com/openshift/configuration-anomaly-detection/blob/e9a58e5d9c3fe4b3c0588fd8ea4a0dbc8041f24b/cadctl/cmd/cluster-missing/cluster-missing.go#L203-L213) is slightly more readable than the [updated version](https://github.com/openshift/configuration-anomaly-detection/blob/436b6c94e1d10592b514e4120501a14b569c2e79/cadctl/cmd/cluster-missing/cluster-missing.go#L210-L240), as this new logic requires deeper nesting and a second check to determine if the cluster was evaluated. I'd also argue that it's not unreasonable to expect the investigation client to have knowledge regarding which failure conditions are retry-able and which are "true" errors, given that the checks it performs are pretty limited in scope.

**Comment by typeid** [comment]
As a heads up: @RaphaelBut  is working on a bigger refactor for CAD that we need for further storys: https://github.com/openshift/configuration-anomaly-detection/pull/152

I think it would be better to add your changes after that PR (which also facilitates retrys). 

I'm also not quite sure why the timeout for a restored cluster is needed, I don't think I've seen an error related to this in the CAD pipelines yet. The current main issue is the node misscount, which is being worked on [here](https://issues.redhat.com/browse/OSD-14629).

**Comment by tnierman** [comment]
> As a heads up: @RaphaelBut is working on a bigger refactor for CAD that we need for further storys: #152
>
> I think it would be better to add your changes after that PR (which also facilitates retrys).

Thanks for letting me know. It looks like that PR is currently held up [waiting for additional unit tests](https://github.com/openshift/configuration-anomaly-detection/pull/152#issuecomment-1420765420), and given that this has potential to impact on-call work, I'm hoping we can merge this first :slightly_smiling_face: 

> I'm also not quite sure why the timeout for a restored cluster is needed, I don't think I've seen an error related to this in the CAD pipelines yet. The current main issue is the node misscount, which is being worked on [here](https://issues.redhat.com/browse/OSD-14629).

The problem this addresses is similar to the node miscount issue: CAD's current evaluation of restored clusters checks if there are non-Running instances belonging to the cluster, failing if it finds any. The original intent, if I remember correctly, was to ensure all 3 master instances were started correctly, since those machines weren't managed by any cluster component at the time and their absence would certainly require SRE attention. 

The actual effect of this check, however, seems to be a disproportionate level of noise for Primary from clusters that have recently terminated machines (ie - if some instances belonging to the cluster hadn't been started with the rest of the cluster, and the machine-api-operator replaced them). Essentially, the criteria being tested is too broad for the case we're trying to catch. 

Removing this check outright will reduce the noise Primary sees from CAD. I don't think it's necessarily worth improving upon either, since 1) for clusters >4.12, the control-plane-machineset-operator should manage missing control-plane nodes like any other machineset (so less need for SRE intervention) and 2) our alerting rules should still notify Primary if the other tests pass, the cluster is taken out of limited support, and a master node is missing. (If the tests fail, then Primary will be alerted anyway).

**Comment by typeid** [comment]
I see, thanks for the details. I'm still unsure how often this occurs, from my point of view we mainly run into the node count bug. 
We investigate the cluster once it checks into DMS again, and I'm assuming that more often than not, the cluster is in a proper state already then. I would think that this PR fixes an edge case, and a high percentage of primary pages come from the node count bug. Please correct me if I'm wrong on that. Also cc @zmird-r @thavlice @RaphaelBut for some input on this.

That said, I do agree we might need to retry the investigation here, but a 20 minute sleep sounds like it's going bring us  the next issues. Would it be better to retry up to e.g. 3 times with a backoff (totalling in up to 20 minutes), so that we can exit the pipeline as soon as possible? 

If this issue really is the main cause for primary disturbance (not node count), then I think we should indeed fix it ASAP .
If not, I think we should rethink how we want to approach handling resolved alerts in general and delay it until after https://github.com/openshift/configuration-anomaly-detection/pull/152. 

We already plan on scheduling a meeting to discuss current CAD issues in general, I will send you an invite as well.

**Comment by typeid** [comment]
>If this issue really is the main cause for primary disturbance (not node count), then I think we should indeed fix it ASAP.

The node count fix is on stage and further highlighted that we also need the changes from this PR urgently for the next CAD promotion. A refactor to _up to 20 minutes_ instead of _after 20 minutes_ will be added in a follow up PR after the refactor https://github.com/openshift/configuration-anomaly-detection/pull/152. 

/lgtm
/approve

**Pull request merged**